cmd/pulumi.go

```diff
@@ -153,3 +161,55 @@ func confirmPrompt(prompt string, name string) bool {
 	line, _ := reader.ReadString('\n')
 	return strings.TrimSpace(line) == name
 }
+
+// reportCommandInvocations hooks into the command's PreRun callback (chaining if necessary) to
+// send a telemetry event indiciating a CLI command was invoked. The command's name is prefixed
+// with namePrefix, so command names are registered as "pulumi-stack-ls".
+func reportCommandInvocations(namePrefix string, cmd *cobra.Command) {
+	// Chain in the PreRun callback.
+	prevPreRun := cmd.PreRun
+	cmd.PreRun = func(cmd *cobra.Command, args []string) {
+		if prevPreRun != nil {
+			prevPreRun(cmd, args)
+		}
+
+		// Walk through the flags that have been set in lexicographical order.
+		var flags []string
+		cmd.Flags().Visit(func(f *pflag.Flag) {
+			flags = append(flags, f.Name)
+		})
+
+		err := reportCommanInvocation(apitype.CommandInvocation{
+			Command: namePrefix,
+			Flags:   flags,
+		})
+		if err != nil {
+			logging.V(7).Infof("error reporting command: %v", err)
+		}
+	}
+
+	// Report child commands if invoked, too.
+	for _, childCmd := range cmd.Commands() {
+		reportCommandInvocations(fmt.Sprintf("%s-%s", namePrefix, childCmd.Name()), childCmd)
+	}
+}
+
+// reportCommanInvocation sends the command invocation event to a web endpoint.
+func reportCommanInvocation(c apitype.CommandInvocation) error {
```

Comment:
This needs to be a backend specific thing, so that fire and forget and disconnected scenarios continue to work.